examples/tls_alpn_provision.rs

```diff
@@ -0,0 +1,363 @@
+//!     --name foo.example.com
+//!
+//! Make sure your DNS has been set up to resolve 'foo.example.com' to the server's IP address.
+//! This will issue a certificate for 'foo.example.com' from the Let's Encrypt staging server.
+//!
+//! You can run this example with Pebble by running:
+//!   cargo run \
+//!     --package instant-acme \
+//!     --example tls_alpn_provision -- \
+//!       --directory-url=https://localhost:14000/dir \
+//!       --server-root-cert-path=<path to pebble CA cert> \
+//!       --port 5001 \
+//!       --name foo.example.com
+//!
+//! Make sure Pebble is running using its default ports and your DNS has been set up to resolve
+//! 'foo.example.com' to localhost (or use Pebble with a mock DNS server like pebble-challtestsrv).
+
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use clap::Parser;
+use hyper_util::client::legacy::Client as HyperClient;
+use hyper_util::rt::TokioExecutor;
+use rcgen::{CertificateParams, CustomExtension, DistinguishedName, KeyPair};
+use rustls::crypto::CryptoProvider;
+use rustls::server::{Acceptor, ClientHello, ResolvesServerCert};
+use rustls::sign::CertifiedKey;
+use rustls::{ClientConfig, RootCertStore, ServerConfig};
+use rustls_pki_types::pem::PemObject;
+use rustls_pki_types::{CertificateDer, PrivatePkcs8KeyDer};
+use rustls_platform_verifier::ConfigVerifierExt;
+use tokio::net::{TcpListener, TcpStream};
+use tokio::sync::mpsc;
+use tokio::time::sleep;
+use tokio::{select, task};
+use tokio_rustls::LazyConfigAcceptor;
+use tracing::{info, warn};
+
+use instant_acme::{
+    Account, AuthorizationStatus, ChallengeType, HttpClient, Identifier, LetsEncrypt, NewAccount,
+    NewOrder, OrderStatus,
+};
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    tracing_subscriber::fmt::init();
+
+    let opts = Options::parse();
+
+    // Build a Rustls config that can validate the ACME server's certificate.
+    let rustls_config = if let Some(roots_pem) = &opts.server_root_cert_path {
+        let mut roots = RootCertStore::empty();
+        roots.add_parsable_certificates(
+            CertificateDer::pem_file_iter(roots_pem)?.map(|result| result.unwrap()),
+        );
+        ClientConfig::builder()
+            .with_root_certificates(roots)
+            .with_no_client_auth()
+    } else {
+        ClientConfig::with_platform_verifier()
+    };
+
+    // Build a Hyper client that can connect to the ACME server.
+    let client = HyperClient::builder(TokioExecutor::new()).build(
+        hyper_rustls::HttpsConnectorBuilder::new()
+            .with_tls_config(rustls_config)
+            .https_or_http()
+            .enable_http1()
+            .enable_http2()
+            .build(),
+    );
+
+    // Spawn a challenge response server that listens for incoming TLS connections to solve
+    // TLS-ALPN-01 challenges.
+    let (shutdown_tx, shutdown_rx) = mpsc::channel(1);
```

Comment:
Maybe use [`CancellationToken`](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html) for this stuff?